_posts/2019-01-10-future-workloads.md

```diff
@@ -0,0 +1,503 @@
+
+Let's look at an example. The following runs the `cpuBound` function 4 times **sequentially** in the main thread.
+
+```
+val time: Int = ???
+(1 to 4).foreach(cpuBound(time))
+```
+
+And using Future instead:
+
+```
+Future.sequence(List.fill(4)(Future(cpuBound(time)))).await()
+```
+
+My approximate simplistic results (in ms):
+
+|           ms| Sequential | Future |
+|-------------|------------|--------|
+| time = 20   | 80         | 173    |
+| time = 40   | 160        | 198    |
+| time = 60   | 240        | 191    |
+| time = 500  | 2000       | 594    |
+| time = 1000 | 4000       | 1103   |
+
+At 20ms, Future is too costly, At 40ms, Future begins to hit the threshold. At 500ms and 1000ms it scales better.
+
+The above is actually a case of [Amdahl's law](https://en.wikipedia.org/wiki/Amdahl%27s_law). The serial part would be the overhead of creating Future in the main thread.
+
+## Blocking
+
+The `blocking` method will produce output similar to the `cpuBound` method, but our cores will not be busy.
+
+```
+Future.sequence(List.fill(1000)(Future(blocking(1000)))).await()
+```
+
+```
+Thread[scala-execution-context-global-11,5,main]
+Thread[scala-execution-context-global-10,5,main]
+Thread[scala-execution-context-global-12,5,main]
+Thread[scala-execution-context-global-13,5,main]
+~1000ms PAUSE
+Thread[scala-execution-context-global-11,5,main]
+Thread[scala-execution-context-global-12,5,main]
+Thread[scala-execution-context-global-10,5,main]
+Thread[scala-execution-context-global-13,5,main]
+... continues until 1000 executions
+```
+
+```
+1  [||||                          8.9%]
+2  [|||                           8.1%]
+3  [|||||                        10.6%]
+4  [||||                          7.3%]
+```
+
+Blocking calls will occupy the threads with no meaningful cpu work. Let's see what happens if we mix blocking and cpu bound workloads and run them on the same thread pool:
+
+```
+def rand = if (Random.nextInt(100) > 50) Future(cpuBound(1000)) else Future(blocking(1000))
+Future.sequence(List.fill(1000)(rand)).await()
+```
+
+Doing this will cause our application to arbitrarily use the cpu. At some point we might have this utilization:
+
+```
+1  [||||                          8.9%]
+2  [|||                           8.1%]
+3  [||||||||||||||||||||||||||||100.0%]
+4  [||||                          7.3%]
+```
+
+As we can see, if we have heterogeneous workload of cpu intensive tasks and blocking, we are actually wasting resources and slowing down our application. A common solution to this problem has been to accept it as is and just schedule all tasks like we did.
+
+But actually, the better solution is to spawn new threads and use those for blocking calls. We can make our application use an extra 4 threads for blocking calls:
```

Comment:
Please remove the "an" before "extra 4 threads", as 4 threads is plural. Also please rephrase to "use 4 extra threads"